bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/CallWrapper.java

```diff
@@ -0,0 +1,93 @@
+package com.google.cloud.bigtable.grpc.async;
+
+import io.grpc.CallOptions;
+import io.grpc.ClientCall;
+import io.grpc.Metadata;
+import io.grpc.stub.ClientCallStreamObserver;
+
+import javax.annotation.Nullable;
+
+public class CallWrapper<RequestT, ResponseT>
+    extends ClientCallStreamObserver<RequestT> {
+  @SuppressWarnings("rawtypes")
+  private static final ClientCall NULL_CALL = new ClientCall() {
+
+    @Override public void start(Listener responseListener, Metadata headers) {
+    }
+
+    @Override public void request(int numMessages) {
+    }
+
+    @Override public void cancel(String message, Throwable cause) {
+    }
+
+    @Override public void halfClose() {
+    }
+
+    @Override public void sendMessage(Object message) {
+    }
+  };
+
+  private boolean autoFlowControlEnabled = true;
+  private ClientCall<RequestT, ResponseT> call = NULL_CALL;
+
+  synchronized void resetCall() {
+    call = NULL_CALL;
```

Comment:
If I understand this correctly. This is intended to be called when we get a transient error. If thats the case, we need to retain the current request count, otherwise the stream will stall:

- something calls request(1)
- transient error occurs
- this gets called, and the request for the message is lost
- no message gets pulled to the listener

Reply:
Those are all good points.  Again, this PR is just a refactor so that we can better focus on issues like the ones you're describing.